src/MetricsTrait.php

```diff
@@ -0,0 +1,124 @@
+<?php
+/*
+ * Copyright 2023 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Auth;
+
+/**
+ * Trait containing helper methods required for enabling
+ * observability metrics in the library.
+ *
+ * @internal
+ */
+trait MetricsTrait
+{
+    /**
+     * @var string The version of the auth library php.
+     */
+    private static $version;
+
+    /**
+     * @var string The header key for the observability metrics.
+     */
+    protected static $metricsHeaderKey = 'x-goog-api-client';
+
+    /**
+     * @var array<string, string> The request type header values
+     *      for the observability metrics.
+     */
+    private static $requestType = [
+        'accessToken' => 'auth-request-type/at',
+        'idToken' => 'auth-request-type/it',
+    ];
+
+    /**
+     * @var array<string, string> The credential type headervalues
+     *      for the observability metrics.
+     */
+    private static $credTypes = [
+        'user' => 'cred-type/u',
+        'sa' => 'cred-type/sa',
+        'jwt' => 'cred-type/jwt',
+        'gce' => 'cred-type/mds',
+        'impersonate' => 'cred-type/imp'
+    ];
+
+    /**
+     * @var string The credential type for the observability metrics.
+     *      This would be overridden by the credential class if applicable.
+     */
+    protected $credType = '';
+
+    protected function getServiceApiMetricsHeaderValue(): string
+    {
+        if (!empty($this->credType)) {
+            return $this->langAndVersion() . ' ' . $this->credType;
+        }
+        return '';
+    }
+
+    protected function getTokenEndpointMetricsHeaderValue(bool $isAccessTokenRequest): string
+    {
+        $value = $this->langAndVersion();
+        if ($isAccessTokenRequest) {
+            $value .= ' ' . self::$requestType['accessToken'];
+        } else {
+            $value .= ' ' . self::$requestType['idToken'];
+        }
+
+        if (!empty($this->credType)) {
+            return $value . ' ' . $this->credType;
+        }
+
+        return '';
+    }
+
+    /**
+     * @param array<mixed> $metadata The metadata to update and return.
+     * @param string $headerValue The header value to add to the metadata for
+     *        observability metrics.
+     * @return array<mixed> The updated metadata.
+     */
+    protected function applyMetricsHeader($metadata, $headerValue)
+    {
```

Comment:
@bshaffer , I was wrong here. On closer inspection, I found that `gax-php` uses headers of the form second whereas `google-cloud-php` and `api-client` uses headers of the form 1st.

Hence we need to necessarily tackle both the cases.

Reply:
We don't need to support merging headers in this library. We can simply set metrics headers if they're not being explicitly passed in. There's no need to support more than that.

Reply:
> We can simply set metrics headers if they're not being explicitly passed in

I'm confused here. Those explicit headers are responsible for tracking metrics for the upstream libraries. We need to set observability metrics for auth library in addition to all those.

Reply:
What I mean by "merge" here is merge a passed-in value of `x-goog-api-client` with _our_ value of `x-goog-api-client`

if I pass in `$metadata` with the value of `x-goog-api-client` already set, we should just use that (or we could override it and say that value can't be set custom, although I'd rather just avoid setting it). But we shouldn't merge both headers as is being done here, we only want to send in one `x-goog-api-client` header.

Reply:
I was referring specifically this part (which you've already removed, so it's no longer relevant)

```php
        } elseif (is_array($metadata[self::$metricsHeaderKey])) {
            $metadata[self::$metricsHeaderKey][0] .= ' ' . $headerValue;
        } else {
            // It's a string instead of array
            $metadata[self::$metricsHeaderKey] .= ' ' . $headerValue;
        }
```

Reply:
Had a internal discussion regarding this and arrived at the conclusion that we do need to merge the `x-goog-api-client` header as it would come autopopulated by the upstream google-gax, google-cloud-php or google-api-client library.

Only in the event when someone tries to use the auth library itself is when `x-goog-api-client` would not be set.